cpp/src/arrow/util/iterator.h

```diff
@@ -28,84 +29,141 @@
 
 namespace arrow {
 
+template <typename T>
+class Iterator;
+
+/// \brief IteratorEnd returns a reserved value which indicates the end of iteration. By
+/// default this is NULLPTR since most iterators yield pointer types. Specialize/overload
+/// if different end semantics are required.
+template <typename T>
+static T IteratorEnd(internal::type_constant<T>) {
+  return T(NULLPTR);
+}
+
+/// Convenience function, overload the type_constant version above and not this one.
+template <typename T>
+static T IteratorEnd() {
+  return IteratorEnd(internal::type_constant<T>{});
+}
+
 /// \brief A generic Iterator that can return errors
 template <typename T>
 class Iterator {
  public:
-  virtual ~Iterator() = default;
+  /// \brief Iterator may be constructed from any type which has member function
+  /// Status HasNext::Next(T*);
+  template <typename HasNext>
+  explicit Iterator(HasNext has_next)
+      : ptr_(new HasNext(std::move(has_next)), Delete<HasNext>), next_(Next<HasNext>) {}
+
+  Iterator() : ptr_(NULLPTR, NoopDelete) {}
 
-  /// \brief Return the next element of the sequence, nullptr when the
-  /// iteration is completed
-  virtual Status Next(T* out) = 0;
+  /// \brief Return the next element of the sequence, IteratorEnd<T>() when the
+  /// iteration is completed. Calling this on a default constructed Iterator
+  /// will result in undefined behavior.
+  Status Next(T* out) { return next_(ptr_.get(), out); }
 
   /// Pass each element of the sequence to a visitor. Will return any error status
   /// returned by the visitor, terminating iteration.
   template <typename Visitor>
   Status Visit(Visitor&& visitor) {
     Status status;
-    T value;
 
-    for (;;) {
+    for (T value, end = IteratorEnd<T>();;) {
       status = Next(&value);
 
       if (!status.ok()) return status;
 
-      if (value == NULLPTR) break;
+      if (value == end) break;
 
       ARROW_RETURN_NOT_OK(visitor(std::move(value)));
     }
 
     return status;
   }
+
+  bool operator==(const Iterator& other) const { return ptr_ == other.ptr_; }
+
+  explicit operator bool() const { return ptr_ == NULLPTR; }
+
+ private:
+  template <typename HasNext>
```

Comment:
Will do

Reply:
Ah, sorry, I wasn't clear. I meant comment on the general design (why we're using type erasure) rather than details of member variables - these can stay of course :-)

Reply:
It can go in another PR though :-)

Reply:
I'll add comments about this in #5221